### PR TITLE
audit: uses_from_macos only uses macOS software

### PR DESCRIPTION
**Pull request by jonchang**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Verify that dependencies specified as being used from macOS are, in fact, provided by macOS.

No tests written because any new test would put `test/dev-cmd/audit_spec.rb` at over 500 lines, an action which is forbidden by Rubocop.

https://github.com/Homebrew/brew/issues/6150
https://github.com/Homebrew/brew/pull/6162

### CODE REVIEW

File: Library/Homebrew/dev-cmd/audit.rb [hunk 1]
```diff
@@ -859,6 +859,17 @@ def line_problems(line, _lineno)
...
       end
 
+      if line =~ /uses_from_macos (["'](.+)["'])/
+        begin
+          macos_f = Formulary.factory(Regexp.last_match(2))
+          unless macos_f.keg_only_reason&.reason == :provided_by_macos
+            problem "`uses_from_macos` should only be used for software provided by macOS."
+          end
+        rescue FormulaUnavailableError
+            problem "`uses_from_macos` dependency #{Regexp.last_match(2)} doesn't exist."
```
Comment (scpeters) [inline Library/Homebrew/dev-cmd/audit.rb]:
```suggestion
          problem "`uses_from_macos` dependency #{Regexp.last_match(2)} doesn't exist."
```

**Comment by MikeMcQuaid** [comment]
> No tests written because any new test would put `test/dev-cmd/audit_spec.rb` at over 500 lines, an action which is forbidden by Rubocop.

Please disable the check or split up the file and write some tests anyway, thanks.

File: Library/Homebrew/dev-cmd/audit.rb [hunk 2]
```diff
@@ -859,6 +859,17 @@ def line_problems(line, _lineno)
         problem "Use `depends_on :java` to set JAVA_HOME"
       end
 
+      if line =~ /uses_from_macos (["'](.+)["'])/
+        begin
+          macos_f = Formulary.factory(Regexp.last_match(2))
+          unless macos_f.keg_only_reason&.reason == :provided_by_macos
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: I'm not sure this is quite true for now; you there's definitely things provided by macOS which aren't keg-only.

File: Library/Homebrew/dev-cmd/audit.rb [hunk 3]
```diff
@@ -859,6 +859,17 @@ def line_problems(line, _lineno)
         problem "Use `depends_on :java` to set JAVA_HOME"
       end
 
+      if line =~ /uses_from_macos (["'](.+)["'])/
```
Comment (sjackman) [inline Library/Homebrew/dev-cmd/audit.rb]:
```suggestion
      if line =~ /uses_from_macos ["']([^"']+)["']/
```

File: Library/Homebrew/dev-cmd/audit.rb [hunk 4]
```diff
@@ -859,6 +859,17 @@ def line_problems(line, _lineno)
...
+        begin
+          macos_f = Formulary.factory(Regexp.last_match(2))
```
Comment (sjackman) [inline Library/Homebrew/dev-cmd/audit.rb]:
```suggestion
          macos_f = Formulary.factory(Regexp.last_match(1))
```

File: Library/Homebrew/dev-cmd/audit.rb (hunk 1 above)
Comment (sjackman) [inline Library/Homebrew/dev-cmd/audit.rb]:
```suggestion
            problem "`uses_from_macos` dependency #{macos_f} doesn't exist."
```

File: Library/Homebrew/dev-cmd/audit.rb (hunk 2 above)
Comment (sjackman) [reply Library/Homebrew/dev-cmd/audit.rb]: True. We could maintain a white list of these formulae, if we wanted to go this way.

File: Library/Homebrew/dev-cmd/audit.rb (hunk 3 above)
Comment (maxim-belkin) [reply Library/Homebrew/dev-cmd/audit.rb]: Can't one capture the first `["']` and then back-reference to it with `\1`?

File: Library/Homebrew/dev-cmd/audit.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/audit.rb]: Yes.

**Comment by sjackman** [comment]
I'm okay with merging this PR and see how it performs in practice, with the understanding that it will likely need tweaking and the creation and maintenance of a white list of formulae provided by macOS in addition to those that are keg-only.

**Comment by MikeMcQuaid** [comment]
> I'm okay with merging this PR and see how it performs in practice, with the understanding that it will likely need tweaking and the creation and maintenance of a white list of formulae provided by macOS in addition to those that are keg-only.

In that case we should add code to handle the whitelist now (even if it's an empty array) and make it a `--strict` audit until we're confident it's reliable enough to be on by default.

**Comment by jonchang** [comment]
> In that case we should add code to handle the whitelist now (even if it's an empty array) and make it a `--strict` audit until we're confident it's reliable enough to be on by default.

Alright, I'll work on this next week.

File: Library/Homebrew/dev-cmd/audit.rb [hunk 5]
```diff
@@ -861,6 +861,24 @@ def line_problems(line, _lineno)
...
+        perl
+        php
+        python@2
+        ruby
+      ].freeze
+
+      if line =~ /uses_from_macos (["'](.+)["'])/
+        begin
+          macos_f = Formulary.factory(Regexp.last_match(2))
+          if macos_f.keg_only_reason&.reason != :provided_by_macos || uses_from_macos_whitelist.exclude?(macos_f.name)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: Can you split this onto two lines? Thanks!

**Comment by jonchang** [comment]
Thanks for the reviews. I'm working on figuring out how to split up the test into multiple files first in a separate pull request, then I will get back to revising this one.

**Comment by MikeMcQuaid** [comment]
> I'm working on figuring out how to split up the test into multiple files first in a separate pull request, then I will get back to revising this one.

I think it's probably worth adding a RuboCop exception for this file instead unless `dev-cmd/audit.rb` itself is split into multiple files.

**Comment by MikeMcQuaid** [comment]
@jonchang any news on this? Thanks!

**Comment by jonchang** [comment]
Haven't figured out how to split up the test file to get it under the 500 line limit. I'll probably try to take another look this weekend.

**Comment by MikeMcQuaid** [comment]
@jonchang Please don't 😄. Either add a `# rubocop:disable` to the relevant file or bump the limits in the `.rubocop.yml`.

**Comment by jonchang** [comment]
Definitely don't have time for this right now. Closing until I go on vacation.

**Pull request closed without merging**